File: src/26_SubstructureInTree/26_SubstructureInTree.cpp
// 面试题26：树的子结构
// 题目：输入两棵二叉树A和B，判断B是不是A的子结构。

/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 * };
 */

#include <iostream>
#include <string>

#include "../Utilities/BinaryTree.h"

using namespace std;

class Solution {
   public:
    bool isSubStructure( TreeNode* A, TreeNode* B ) {
        // 分两步：1) 遍历A的每个节点nA，2) 判断以nA为根节点的子树是否包含B
        // 先以A为根节点，如果找不到B，则继续在左右子树寻找
        if ( !A || !B ) return false;
        return isAHasB( A, B ) || isSubStructure( A->left, B ) ||
               isSubStructure( A->right, B );
    }

   private:
    // 递归判断A是否包含B
    bool isAHasB( TreeNode* A, TreeNode* B ) {
        // 成功条件：已经遍历至B的叶节点
        if ( !B ) return true;
        // 失败条件：1) 到A的叶节点，但没到B的叶节点; 2) 当前节点值不相同
        if ( !A || A->val != B->val ) return false;
        // 继续遍历左右子树
        return isAHasB( A->left, B->left ) && isAHasB( A->right, B->right );
    }
};

class Test {
   public:
    void checkResult( const string testName, TreeNode* A, TreeNode* B,
                      bool expected ) {
        cout << testName;
        bool result = solution.isSubStructure( A, B );
        if ( result == expected )
            cout << " passed.\n";
        else
            cout << " failed.\n";
    }

    // 树中结点含有分叉，树B是树A的子结构
    //                  8                8
    //              /       \           / \
    //             8         7         9   2
    //           /   \
    //          9     2
    //               / \
    //              4   7
    void Test1() {
        ValVec Avals{ 8, 8, 7, 9, 2, 4, 7 };
        ValVec Bvals{ 8, 9, 2 };

        TreeNodeVec Anodes = createTreeNodesVec( Avals );
        TreeNodeVec Bnodes = createTreeNodesVec( Bvals );

        TreeIndexVec Aidxs{
            { 0, 1, 2 },
            { 1, 3, 4 },
            { 4, 5, 6 },
        };
        TreeIndexVec Bidxs{
            { 0, 1, 2 },
        };
        connectTreeNodesIndexVec( Anodes, Aidxs );
        connectTreeNodesIndexVec( Bnodes, Bidxs );

        checkResult( "Test1", Anodes[ 0 ], Bnodes[ 0 ], true );
        destroyTree( Anodes[ 0 ] );
        destroyTree( Bnodes[ 0 ] );
    }

    // 树中结点含有分叉，树B不是树A的子结构
    //                  8                8
    //              /       \           / \
    //             8         7         9   2
    //           /   \
    //          9     3
    //               / \
    //              4   7
    void Test2() {
        ValVec Avals{ 8, 8, 7, 9, 3, 4, 7 };
        ValVec Bvals{ 8, 9, 2 };

        TreeNodeVec Anodes = createTreeNodesVec( Avals );
        TreeNodeVec Bnodes = createTreeNodesVec( Bvals );

        TreeIndexVec Aidxs{
            { 0, 1, 2 },
            { 1, 3, 4 },
            { 4, 5, 6 },
        };
        TreeIndexVec Bidxs{
            { 0, 1, 2 },
        };
        connectTreeNodesIndexVec( Anodes, Aidxs );
        connectTreeNodesIndexVec( Bnodes, Bidxs );

        checkResult( "Test2", Anodes[ 0 ], Bnodes[ 0 ], false );
        destroyTree( Anodes[ 0 ] );
        destroyTree( Bnodes[ 0 ] );
    }

    // 树中结点只有左子结点，树B是树A的子结构
    //                8                  8
    //              /                   /
    //             8                   9
    //           /                    /
    //          9                    2
    //         /
    //        2
    //       /
    //      5
    void Test3() {
        ValVec Avals{ 8, 8, 9, 2, 5 };
        ValVec Bvals{ 8, 9, 2 };

        TreeNodeVec Anodes = createTreeNodesVec( Avals );
        TreeNodeVec Bnodes = createTreeNodesVec( Bvals );

        TreeIndexVec Aidxs{
            { 0, 1, -1 },
            { 1, 2, -1 },
            { 2, 3, -1 },
            { 3, 4, -1 },
        };
        TreeIndexVec Bidxs{
            { 0, 1, -1 },
            { 1, 2, -1 },
        };
        connectTreeNodesIndexVec( Anodes, Aidxs );
        connectTreeNodesIndexVec( Bnodes, Bidxs );

        checkResult( "Test3", Anodes[ 0 ], Bnodes[ 0 ], true );
        destroyTree( Anodes[ 0 ] );
        destroyTree( Bnodes[ 0 ] );
    }

    // 树中结点只有左子结点，树B不是树A的子结构
    //                8                  8
    //              /                   /
    //             8                   9
    //           /                    /
    //          9                    3
    //         /
    //        2
    //       /
    //      5
    void Test4() {
        ValVec Avals{ 8, 8, 9, 2, 5 };
        ValVec Bvals{ 8, 9, 3 };

        TreeNodeVec Anodes = createTreeNodesVec( Avals );
        TreeNodeVec Bnodes = createTreeNodesVec( Bvals );

        TreeIndexVec Aidxs{
            { 0, 1, -1 },
            { 1, 2, -1 },
            { 2, 3, -1 },
            { 3, 4, -1 },
        };
        TreeIndexVec Bidxs{
            { 0, 1, -1 },
            { 1, 2, -1 },
        };
        connectTreeNodesIndexVec( Anodes, Aidxs );
        connectTreeNodesIndexVec( Bnodes, Bidxs );

        checkResult( "Test4", Anodes[ 0 ], Bnodes[ 0 ], false );
        destroyTree( Anodes[ 0 ] );
        destroyTree( Bnodes[ 0 ] );
    }

    // 树中结点只有右子结点，树B是树A的子结构
    //       8                   8
    //        \                   \ 
    //         8                   9
    //          \                   \
    //           9                   2
    //            \      
    //             2
    //              \
    //               5
    void Test5() {
        ValVec Avals{ 8, 8, 9, 2, 5 };
        ValVec Bvals{ 8, 9, 2 };

        TreeNodeVec Anodes = createTreeNodesVec( Avals );
        TreeNodeVec Bnodes = createTreeNodesVec( Bvals );

        TreeIndexVec Aidxs{
            { 0, -1, 1 },
            { 1, -1, 2 },
            { 2, -1, 3 },
            { 3, -1, 4 },
        };
        TreeIndexVec Bidxs{
            { 0, -1, 1 },
            { 1, -1, 2 },
        };
        connectTreeNodesIndexVec( Anodes, Aidxs );
        connectTreeNodesIndexVec( Bnodes, Bidxs );

        checkResult( "Test5", Anodes[ 0 ], Bnodes[ 0 ], true );
        destroyTree( Anodes[ 0 ] );
        destroyTree( Bnodes[ 0 ] );
    }

    // 树中结点只有右子结点，树B不是树A的子结构
    //       8                   8
    //        \                   \ 
    //         8                   9
    //          \                 / \
    //           9               3   2
    //            \      
    //             2
    //              \
    //               5
    void Test6() {
        ValVec Avals{ 8, 8, 9, 2, 5 };
        ValVec Bvals{ 8, 9, 3, 2 };

        TreeNodeVec Anodes = createTreeNodesVec( Avals );
        TreeNodeVec Bnodes = createTreeNodesVec( Bvals );

        TreeIndexVec Aidxs{
            { 0, -1, 1 },
            { 1, -1, 2 },
            { 2, -1, 3 },
            { 3, -1, 4 },
        };
        TreeIndexVec Bidxs{
            { 0, -1, 1 },
            { 1, 2, 3 },
        };
        connectTreeNodesIndexVec( Anodes, Aidxs );
        connectTreeNodesIndexVec( Bnodes, Bidxs );

        checkResult( "Test6", Anodes[ 0 ], Bnodes[ 0 ], false );
        destroyTree( Anodes[ 0 ] );
        destroyTree( Bnodes[ 0 ] );
    }

    // 树A为空树
    void Test7() {
        ValVec Bvals{ 8, 9, 3, 2 };
        TreeNodeVec Bnodes = createTreeNodesVec( Bvals );

        TreeIndexVec Bidxs{
            { 0, -1, 1 },
            { 1, 2, 3 },
        };
        connectTreeNodesIndexVec( Bnodes, Bidxs );

        checkResult( "Test7", nullptr, Bnodes[ 0 ], false );
        destroyTree( Bnodes[ 0 ] );
    }

    // 树B为空树
    void Test8() {
        ValVec Avals{ 8, 8, 9, 2, 5 };
        TreeNodeVec Anodes = createTreeNodesVec( Avals );

        TreeIndexVec Aidxs{
            { 0, -1, 1 },
            { 1, -1, 2 },
            { 2, -1, 3 },
            { 3, -1, 4 },
        };
        connectTreeNodesIndexVec( Anodes, Aidxs );

        checkResult( "Test8", Anodes[ 0 ], nullptr, false );
        destroyTree( Anodes[ 0 ] );
    }

    // 树A和树B都为空
    void Test9() { checkResult( "Test9", nullptr, nullptr, false ); }

    void runTest() {
        Test1();
        Test2();
        Test3();
        Test4();
        Test5();
        Test6();
        Test7();
        Test8();
        Test9();
    }

   private:
    Solution solution;
};

int main() {
    Test test;
    test.runTest();
    return 0;
}

File: src/26_SubstructureInTree/26_SubstructureInTree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None


import os
import sys

sys.path.append(os.path.split(sys.path[0])[0])


from Utilities import binary_tree
from Utilities.binary_tree import TreeNode


class Solution:
    def isSubStructure(self, A: TreeNode, B: TreeNode) -> bool:
        if not A or not B:
            return False
        return (
            self.isAHasB(A, B)
            or self.isSubStructure(A.left, B)
            or self.isSubStructure(A.right, B)
        )

    def isAHasB(self, A: TreeNode, B: TreeNode) -> bool:
        if not B:
            return True
        if not A or A.val != B.val:
            return False
        return self.isAHasB(A.left, B.left) and self.isAHasB(A.right, B.right)


class Test:
    def __init__(self) -> None:
        self.solution = Solution()

    def checkResult(self, testName: str, A: TreeNode, B: TreeNode, expected: bool):
        text = testName
        result = self.solution.isSubStructure(A, B)
        if result == expected:
            text += " passed."
        else:
            text += " failed."
        print(text)

    # 树中结点含有分叉，树B是树A的子结构
    #                  8                8
    #              /       \           / \
    #             8         7         9   2
    #           /   \
    #          9     2
    #               / \
    #              4   7
    def Test1(self):
        Avals = [8, 8, 7, 9, 2, 4, 7]
        Bvals = [8, 9, 2]

        Anodes = binary_tree.createTreeNodesVec(Avals)
        Bnodes = binary_tree.createTreeNodesVec(Bvals)

        Aidxs = [
            (0, 1, 2),
            (1, 3, 4),
            (4, 5, 6),
        ]
        Bidxs = [(0, 1, 2)]
        binary_tree.connectTreeNodesIndexVec(Anodes, Aidxs)
        binary_tree.connectTreeNodesIndexVec(Bnodes, Bidxs)

        self.checkResult("Test1", Anodes[0], Bnodes[0], True)

    # 树中结点含有分叉，树B不是树A的子结构
    #                  8                8
    #              /       \           / \
    #             8         7         9   2
    #           /   \
    #          9     3
    #               / \
    #              4   7
    def Test2(self):
        Avals = [8, 8, 7, 9, 3, 4, 7]
        Bvals = [8, 9, 2]

        Anodes = binary_tree.createTreeNodesVec(Avals)
        Bnodes = binary_tree.createTreeNodesVec(Bvals)

        Aidxs = [
            (0, 1, 2),
            (1, 3, 4),
            (4, 5, 6),
        ]
        Bidxs = [(0, 1, 2)]
        binary_tree.connectTreeNodesIndexVec(Anodes, Aidxs)
        binary_tree.connectTreeNodesIndexVec(Bnodes, Bidxs)

        self.checkResult("Test2", Anodes[0], Bnodes[0], False)

    # 树中结点只有左子结点，树B是树A的子结构
    #                8                  8
    #              /                   /
    #             8                   9
    #           /                    /
    #          9                    2
    #         /
    #        2
    #       /
    #      5
    def Test3(self):
        Avals = [8, 8, 9, 2, 5]
        Bvals = [8, 9, 2]

        Anodes = binary_tree.createTreeNodesVec(Avals)
        Bnodes = binary_tree.createTreeNodesVec(Bvals)

        Aidxs = [
            (0, 1, -1),
            (1, 2, -1),
            (2, 3, -1),
            (3, 4, -1),
        ]
        Bidxs = [
            (0, 1, -1),
            (1, 2, -1),
        ]
        binary_tree.connectTreeNodesIndexVec(Anodes, Aidxs)
        binary_tree.connectTreeNodesIndexVec(Bnodes, Bidxs)

        self.checkResult("Test3", Anodes[0], Bnodes[0], True)

    # 树中结点只有左子结点，树B不是树A的子结构
    #                8                  8
    #              /                   /
    #             8                   9
    #           /                    /
    #          9                    3
    #         /
    #        2
    #       /
    #      5
    def Test4(self):
        Avals = [8, 8, 9, 2, 5]
        Bvals = [8, 9, 3]

        Anodes = binary_tree.createTreeNodesVec(Avals)
        Bnodes = binary_tree.createTreeNodesVec(Bvals)

        Aidxs = [
            (0, 1, -1),
            (1, 2, -1),
            (2, 3, -1),
            (3, 4, -1),
        ]
        Bidxs = [
            (0, 1, -1),
            (1, 2, -1),
        ]
        binary_tree.connectTreeNodesIndexVec(Anodes, Aidxs)
        binary_tree.connectTreeNodesIndexVec(Bnodes, Bidxs)

        self.checkResult("Test4", Anodes[0], Bnodes[0], False)

    # 树中结点只有右子结点，树B是树A的子结构
    #       8                   8
    #        \                   \
    #         8                   9
    #          \                   \
    #           9                   2
    #            \
    #             2
    #              \
    #               5
    def Test5(self):
        Avals = [8, 8, 9, 2, 5]
        Bvals = [8, 9, 2]

        Anodes = binary_tree.createTreeNodesVec(Avals)
        Bnodes = binary_tree.createTreeNodesVec(Bvals)

        Aidxs = [
            (0, -1, 1),
            (1, -1, 2),
            (2, -1, 3),
            (3, -1, 4),
        ]
        Bidxs = [
            (0, -1, 1),
            (1, -1, 2),
        ]
        binary_tree.connectTreeNodesIndexVec(Anodes, Aidxs)
        binary_tree.connectTreeNodesIndexVec(Bnodes, Bidxs)

        self.checkResult("Test5", Anodes[0], Bnodes[0], True)

    # 树中结点只有右子结点，树B不是树A的子结构
    #       8                   8
    #        \                   \
    #         8                   9
    #          \                 / \
    #           9               3   2
    #            \
    #             2
    #              \
    #               5
    def Test6(self):
        Avals = [8, 8, 9, 2, 5]
        Bvals = [8, 9, 3, 2]

        Anodes = binary_tree.createTreeNodesVec(Avals)
        Bnodes = binary_tree.createTreeNodesVec(Bvals)

        Aidxs = [
            (0, -1, 1),
            (1, -1, 2),
            (2, -1, 3),
            (3, -1, 4),
        ]
        Bidxs = [
            (0, -1, 1),
            (1, 2, 3),
        ]
        binary_tree.connectTreeNodesIndexVec(Anodes, Aidxs)
        binary_tree.connectTreeNodesIndexVec(Bnodes, Bidxs)

        self.checkResult("Test6", Anodes[0], Bnodes[0], False)

    # 树A为空树
    def Test7(self):
        Bvals = [8, 9, 3, 2]
        Bnodes = binary_tree.createTreeNodesVec(Bvals)

        Bidxs = [
            (0, -1, 1),
            (1, 2, 3),
        ]
        binary_tree.connectTreeNodesIndexVec(Bnodes, Bidxs)

        self.checkResult("Test7", None, Bnodes[0], False)

    # 树B为空树
    def Test8(self):
        Avals = [8, 8, 9, 2, 5]
        Anodes = binary_tree.createTreeNodesVec(Avals)

        Aidxs = [
            (0, -1, 1),
            (1, -1, 2),
            (2, -1, 3),
            (3, -1, 4),
        ]
        binary_tree.connectTreeNodesIndexVec(Anodes, Aidxs)

        self.checkResult("Test8", Anodes[0], None, False)

    # 树A和树B都为空
    def Test9(self):
        self.checkResult("Test9", None, None, False)

    def runTest(self):
        self.Test1()
        self.Test2()
        self.Test3()
        self.Test4()
        self.Test5()
        self.Test6()
        self.Test7()
        self.Test8()
        self.Test9()


if __name__ == "__main__":
    test = Test()
    test.runTest()
